### PR TITLE
Line of sight and collision fixes

### LINKED ISSUE #314: [Bug]: Arrows can fly over walls and through roofs and hit pawns through solid objects.
**Complete the template below. Reports that do not follow the template will be deleted.**
**Title should be:** [Bug]: Arrows can fly over walls and through roofs and hit pawns through solid objects.

**Specifications**
Please provide the following.

Rimworld version: 1.2.2753
Combat Extended version: 3.2 (Latest) as of maybe a day or two ago.
Combat Extended source (Steam, GitHub, etc.): I downloaded it directly from this github.
Your operating system: Windows 10
Your mod list: https://gist.github.com/HugsLibRecordKeeper/cf1604ce574080fc524bc870f8b6fc85

**Description**
I had a pawn shooting at some bulls through an embrasure with arrows. The pawn missed, and hit another pawn who is out of line of sight, and sleeping inside the base. There is a roof and solid wall in the way, but the arrow went through one or the other to hit the pawn in the head. Perhaps line of sight rules should apply to missed projectiles too?

**Expected behavior**
The roof or the wall would have stopped the arrow.

**To reproduce**
Have a pawn shooting arrows at a different creature with a wall/roof structure near by, and the arrows may hit things protected by the roof and wall combo even though line of sight is false.

**Screenshots & log dumps**
https://imgur.com/a/1I8IqQg

**Complete the following checklist**
I hereby verify that I have done the following:
- [ ] Confirmed that my game version and load order are correct.
- [ ] Confirmed that I am running the appropriate and most updated version of Combat Extended and required compatibility patches.
- [ ] Confirmed I am not running any mods with known incompatibilities with Combat Extended.
- [ ] Disabled Combat Extended and attempted to reproduce the behavior without success.

### LINKED ISSUE #500: [Bug]: Always miss and hit embrasure when shooting through corner embrasure
**Specifications**

Rimworld version:1.2.2753
Combat Extended version: 1.2 (December 7 update from steam)
Combat Extended source (Steam, GitHub, etc.): Steam
Your operating system: Mac
Your mod list:

<details>
<summary>Mod List</summary>

Version: 1
Name: bug
Mods:
- Id: brrainz.harmony
  Name: Harmony
- Id: ludeon.rimworld
  Name: Core
- Id: ludeon.rimworld.royalty
  Name: Royalty
- Id: unlimitedhugs.hugslib
  Name: HugsLib
- Id: fluffy.modmanager
  Name: Mod Manager
- Id: madeline.modmismatchformatter
  Name: Better ModMismatch Window
- Id: edb.preparecarefully
  Name: EdB Prepare Carefully
- Id: rensen.msoresearch
  Name: More Scenario Options - Research
- Id: arandomkiwi.rimsaves
  Name: RimSaves
- Id: vanillaexpanded.vtexe
  Name: Vanilla Textures Expanded
- Id: telefonmast.graphicssettings
  Name: Graphics Settings+
- Id: nandonalt.snowytrees
  Name: Snowy Trees
- Id: sarg.smartspeed
  Name: Smart Speed
- Id: fluffy.followme
  Name: Follow Me
- Id: saucypigeon.showfertility
  Name: Show Fertility
- Id: falconne.heatmap
  Name: Heat Map
- Id: krafs.levelup
  Name: Level Up!
- Id: uuugggg.listeverything
  Name: List Everything
- Id: mehni.numbers
  Name: Numbers
- Id: bodlosh.weaponstats
  Name: WeaponStats
- Id: dubwise.dubsmintmenus
  Name: Dubs Mint Menus
- Id: nals.dynamicportraits
  Name: '[NL] Dynamic Portraits'
- Id: jaxe.rimhud
  Name: RimHUD
- Id: lostrd.wildanimalalert
  Name: Wild Animal Alert
- Id: gguake.ui.simplesearchbar
  Name: Simple Search Bar
- Id: navyseal5.moreharvestdesignators
  Name: More Harvest Designators!
- Id: erdelf.exportagency
  Name: Export Agency
- Id: notfood.researchpal
  Name: ResearchPal
- Id: tammybee.raiderinfo
  Name: RaiderInfo
- Id: fluffy.relationstab
  Name: Relations Tab
- Id: fluffy.medicaltab
  Name: Medical Tab
- Id: unlimitedhugs.allowtool
  Name: Allow Tool
- Id: hatti.qualitybuilder
  Name: QualityBuilder
- Id: longwater.smartstockpilecreation
  Name: Quick Stockpile Creation
- Id: uuugggg.whatismypurpose
  Name: What Is My Purpose
- Id: mlph.openeddoorsdontblocklight
  Name: Opened Doors Don't Block Light
- Id: uuugggg.everybodygetsone
  Name: Everybody Gets One
- Id: mlie.pickupandhaul
  Name: Pick Up And Haul (Continued)
- Id: uuugggg.smartmedicine
  Name: Smart Medicine
- Id: uuugggg.usebedrolls
  Name: Use Bedrolls
- Id: tradingspot.kv.rw
  Name: '[KV] Trading Spot'
- Id: mlie.dismisstrader
  Name: Dismiss Trader (Continued)
- Id: fyarn.priorityclean
  Name: PriorityClean
- Id: eagle0600.dresspatients
  Name: Dress Patients
- Id: fluffy.colonymanager
  Name: Colony Manager
- Id: frozensnowfox.complexjobs
  Name: '[FSF] Complex Jobs'
- Id: uuugggg.randomresearch
  Name: Random Research
- Id: ceteam.combatextended
  Name: Combat Extended
- Id: petetimessix.simplesidearms
  Name: Simple sidearms
- Id: nykot.rpgstyleinventoryce
  Name: RPG Style Inventory CE Revived [1.2]
- Id: voult.betterpawncontrol
  Name: Better Pawn Control
- Id: ceteam.combatextendedguns
  Name: Combat Extended Guns
- Id: n7huntsman.combatextendedshields
  Name: 'Combat Extended: Shields'
- Id: showhair.kv.rw
  Name: '[KV] Show Hair With Hats or Hide All Hats'
- Id: mireia.bodies
  Name: Beautiful Bodies
- Id: children.and.pregnancy
  Name: Children and Pregnancy
- Id: roolo.runandgun
  Name: RunAndGun
- Id: roolo.giddyupcore
  Name: Giddy-up! Core
- Id: roolo.giddyupbattlemounts
  Name: Giddy-up! Battle Mounts
- Id: roolo.giddyupcaravan
  Name: Giddy-up! Caravan
- Id: roolo.giddyuprideandroll
  Name: Giddy-up! Ride and Roll
- Id: spoonshortage.adogsaidanimalprosthetics
  Name: A Dog Said... Animal Prosthetics
- Id: lbeaston.adogsaideasypatcher
  Name: LBE's A Dog Said Easy Patcher
- Id: vanillaexpanded.vaeas
  Name: Vanilla Animals Expanded — Arid Shrubland
- Id: vanillaexpanded.vaeau
  Name: Vanilla Animals Expanded — Australia
- Id: vanillaexpanded.vaet
  Name: Vanilla Animals Expanded — Tundra
- Id: vanillaexpanded.vaets
  Name: Vanilla Animals Expanded — Tropical Swamp
- Id: vanillaexpanded.vaetr
  Name: Vanilla Animals Expanded — Tropical Rainforest
- Id: vanillaexpanded.vaeis
  Name: Vanilla Animals Expanded — Ice Sheet
- Id: vanillaexpanded.vaeed
  Name: Vanilla Animals Expanded — Extreme Desert
- Id: vanillaexpanded.vaebf
  Name: Vanilla Animals Expanded — Boreal Forest
- Id: vanillaexpanded.vaecd
  Name: Vanilla Animals Expanded — Cats and Dogs
- Id: vanillaexpanded.vaed
  Name: Vanilla Animals Expanded — Desert
- Id: vanillaexpanded.vaeendandext
  Name: Vanilla Animals Expanded — Endangered
- Id: oskarpotocki.vanillafactionsexpanded.core
  Name: Vanilla Expanded Framework
- Id: vanillaexpanded.vaetf
  Name: Vanilla Animals Expanded — Temperate Forest
- Id: jamaicancastle.rf.fertilefields
  Name: Fertile Fields [1.1]
- Id: notfood.seedsplease
  Name: SeedsPlease
- Id: shadow.spaiomegapatch
  Name: 'SeedsPlease: All-In-One Megapatch'
- Id: dracoix.doormat.r12a
  Name: Door Mat
- Id: earthyturtle.letstrade
  Name: Let's Trade!
- Id: peppsen.pmusic
  Name: P-Music
- Id: wemd.realisticdarkness
  Name: '[WD] Realistic Darkness'

</details>

**Description**
I tried to shoot a raider with a short bow through an embrasure. It consistently narrowly misses, hitting the embrasure instead.

**Expected behavior**
I expect the arrow to hit the raider often. I am only two squares away. If I move one square over, I hit the raider consistently. Keep in mind in both cases there is an embrasure between me and the raider. But in the first case, I always miss, and in the second case, I almost always hit.

**To reproduce**

The arrangement should be exactly like so:

W = wall
E = embrasure
C = colonist
R = Raider
X = Wall, embrasure, or door. actively being attacked by the raider

<pre>
          R
W W W E X E            < - - - Cannot hit raider!
        C E 
          W W W W
</pre>

My weapon is a short bow with stone arrows and my shooting skill is around 5.

Now, in this position, I miss every single time. The log says "narrowly missed, but hit embrasure instead".

Now say we rearrange things a bit:

<pre>
        R
W W W E X E          < - - - CAN hit raider!         
        C E 
          W W W W

          R
W W W E X E          < - - - CAN hit raider!         
      C   E 
          W W W W

          
W W W E E E          < - - - CAN hit raider!         
      C   E R 
          W X W W

W W W E E E          < - - - CAN hit raider!         
        C E R 
          W X W W
</pre>

In all the cases above where I said "CAN hit raider", the hit rate is very high, near 100%. However in the specific position,
 I showed first, the hit rate is 0.

**Screenshots & log dumps**

Completely impossible to hit raider like this
Screen Shot 2021-01-25 at 9 13 40 PM

mouseover tooltip in problematic position
Screen Shot 2021-01-25 at 9 14 52 PM

mouseover tooltip in position with no problems (near 100% accuracy)
Screen Shot 2021-01-25 at 9 15 28 PM

**Complete the following checklist**
I hereby verify that I have done the following:
- [X] Confirmed that my game version and load order are correct.
- [X] Confirmed that I am running the appropriate and most updated version of Combat Extended and required compatibility patches.
- [X] Confirmed I am not running any mods with known incompatibilities with Combat Extended.
- [N/A] Disabled Combat Extended and attempted to reproduce the behavior without success.

### LINKED ISSUE #581: [Discussion]: Pawns keep shooting the person right in front of them
In vanilla, when pawns are only 1 tile away from your colonist's standing position, they don't shoot them. It makes sense, as realistically, either the person in front of them would be ducking or the one standing behind stands slightly to the side while shooting, using the person in front of them as a cover. It frustrates me every time insects pop up and I put my melee colonists in front reflexively, only to have the person behind them shoot them right in the head. It makes me unable to play permadeath.

I even downloaded the "Avoid Friendly Fire" mod to avoid them shooting when someone is in front of them but that mod also acknowledges the pawn in front of the colonist has no chance of getting hit, so it won't prevent them from shooting, so my colonists still shoot my melee guys right in the head from 2 centimeters behind them.

I know this is probably an issue with the new projectile based system but couldn't there be a possible adjustment for this? Like making the friendly pawns in front count as a wall to take cover behind somehow? Or I noticed that the colonists sometimes duck when they're behind cover, make the melee pawn do that when in this situation? I probably want too much but I just want my melee colonists to be useful without getting constantly assassinated by my own dudes.

### LINKED ISSUE #597: [Bug]: No projectile collisions on last tick
Projectiles don't collide with objects on their last tick, causing high-velocity projectiles that can move more than 1 cell per tick to sometimes miss targets directly in their path.

**To reproduce**
- Have a heavy charge blaster centipede (projectile speed 180 in current CE version) fire at a target 1 cell away vertically, and 3 cells away horizontally
image

### PR DESCRIPTION
**Pull request by LockdownX7**
## Changes
- Prevent pawns from firing at distant enemies when there's a friendly partially obstructing the target
- Replace line of sight calculations with more reliable system (fixes #581)
- Fix pawns sometimes able to lean to one side of a wall but not another, when firing at targets parallel to the wall
- Update collision skip checks, fixing edge cases where a projectile collides with an object it shouldn't, or phases through an obstruction that should have stopped it (fixes #500, fixes #314)
- Projectiles now check for collisions on their last tick (fixes #597)

## Testing

Check tests you have performed:
- [x] Compiles without warnings
- [x] Game runs without errors
- [x] Playtested a colony (2 days)

### CODE REVIEW

File: Source/CombatExtended/CombatExtended/Verbs/Verb_LaunchProjectileCE.cs [hunk 1]
```diff
@@ -286,11 +286,11 @@ protected virtual void ShiftTarget(ShiftVecReport report, bool calculateMechanic
...
-                                                                /*if (currentTarget.Thing is Building && CompFireModes?.CurrentAimMode == AimMode.SuppressFire)
-                                                                {
-                                                                    targetRange.min = targetRange.max;
-                                                                    targetRange.max = targetRange.min + 1f;
-                                                                }*/
+                    /*if (currentTarget.Thing is Building && CompFireModes?.CurrentAimMode == AimMode.SuppressFire)
+                    {
+                        targetRange.min = targetRange.max;
+                        targetRange.max = targetRange.min + 1f;
+                    }*/
```
Comment (MadaraUchiha) [inline Source/CombatExtended/CombatExtended/Verbs/Verb_LaunchProjectileCE.cs:293]: Consider removing commented code altogether.

File: Source/CombatExtended/CombatExtended/Verbs/Verb_LaunchProjectileCE.cs [hunk 2]
```diff
@@ -565,10 +565,8 @@ protected override bool TryCastShot()
                 //New aiming algorithm
                 projectile.canTargetSelf = false;
 
-                var targDist = (sourceLoc - currentTarget.Cell.ToIntVec2.ToVector2Shifted()).magnitude;
-                if (targDist <= 2)
-                    targDist *= 2;  // Double to account for divide by 4 in ProjectileCE minimum collision distance calculations
-                projectile.minCollisionSqr = Mathf.Pow(targDist, 2);
+                var targetDistance = (sourceLoc - currentTarget.Cell.ToIntVec2.ToVector2Shifted()).magnitude;
+                projectile.minCollisionDistance = Mathf.Min(1.5f, targetDistance * 0.75f);
```
Comment (MadaraUchiha) [inline Source/CombatExtended/CombatExtended/Verbs/Verb_LaunchProjectileCE.cs:569]: I like this 👍

**Review by MadaraUchiha — APPROVED** [review]
LGTM, one small nitpick to consider, but overall OK to merge IMO.

**Review by N7Huntsman — DISMISSED** [review]
No issues encountered during playtest, though I didn't rigorously test for the changes in collision behavior.

**Review by N7Huntsman — APPROVED** [review]

**Pull request merged**